tchannel/zipkin/tracers.py

```diff
@@ -135,23 +139,23 @@ def __init__(self, tchannel):
             A tchannel instance to send the trace info to zipkin server
         """
 
-        from ..thrift import client_for
-
-        # TODO add different thrift service name
-        TCollectorClient = client_for('tcollector', TCollector)
-
         self._tchannel = tchannel
-        self.client = TCollectorClient(self._tchannel)
 
-    def record(self, traces):
-        def submit_callback(f):
-            if f.exception():
-                log.error('Fail to submit zipkin trace',
-                          exc_info=f.exc_info())
+    def submit_callback(self, f):
+        if f.exception():
+            log.error(
+                'Fail to submit zipkin trace',
+                exc_info=f.exc_info()
+            )
 
+    def record(self, traces):
         for (trace, annotations) in traces:
-            f = self.client.submit(thrift_formatter(trace, annotations))
-            f.add_done_callback(submit_callback)
+            client = TCollectorClient(
+                self._tchannel,
+                protocol_headers={'shardKey': i64_to_base64(trace.trace_id)}
+            )
+            f = client.submit(thrift_formatter(trace, annotations))
+            f.add_done_callback(self.submit_callback)
```

Comment:
This is good, but you want to move submit_callback back. It's okay if it's a lambda or a `_` method. Basically, it's not part of the public interface.